scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,188 @@
+        log_softmax = torch.nn.LogSoftmax()
+        theta_log = log_softmax(self.theta_logit)  # (5)
+
+        # compute mean of NegBin - shape (n_cells, n_genes, n_labels)
+        s = x.sum(1, keepdim=True)  # (128, 1)
+        base_mean = torch.log(s)
+        base_mean_u = base_mean.unsqueeze(-1)  # (128, 1, 1)
+        base_mean_e = base_mean_u.expand(
+            s.shape[0], self.n_genes, self.n_labels
+        )  # (128, 100, 5)
+
+        delta_rho = delta * self.rho
+        delta_rho_e = delta_rho.expand(
+            s.shape[0], self.n_genes, self.n_labels
+        )  # (128, 100, 5)
+        log_mu_ngc = base_mean_e + delta_rho_e
+        mu_ngc = torch.exp(log_mu_ngc)  # (128, 100, 5)
+
+        # compute basis means for phi - shape (B)
+        basis_means_fixed = np.linspace(torch.min(x), torch.max(x), self.B)
+        basis_means = torch.tensor(basis_means_fixed)  # (10)
+
+        # compute phi of NegBin - shape (n_cells, n_genes, n_labels)
+        a = torch.exp(self.log_a)  # (10)
+        a_e = a.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        b_init = 2 * ((basis_means_fixed[1] - basis_means_fixed[0]) ** 2)
+        b = torch.exp(torch.ones(self.B) * (-np.log(b_init)))  # (10)
+        b_e = b.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        mu_ngc_u = mu_ngc.unsqueeze(-1)
+        mu_ngcb = mu_ngc_u.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (128, 100, 5, 10)
+        basis_means_e = basis_means.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (128, 100, 5, 10)
+        phi = (  # (128, 100, 5)
+            torch.sum(a_e * torch.exp(-b_e * torch.square(mu_ngcb - basis_means_e)), 3)
+            + self.LOWER_BOUND
+        )
+
+        # compute gamma
+        pdb.set_trace()
+        nb_pdf = NegativeBinomial(probs=mu_ngc, total_count=phi)
+        y_ = x.unsqueeze(-1).expand(s.shape[0], self.n_genes, self.n_labels)
+        y_log_prob_raw = nb_pdf.log_prob(y_)  # (128, 100, 5)
+        theta_log_e = theta_log.expand(128, 5)
+        p_y_unorm = torch.sum(y_log_prob_raw, 1) + theta_log_e  # (128, 5)
+        p_y_norm = torch.logsumexp(p_y_unorm, 1)
+        p_y_norm_e = p_y_norm.unsqueeze(-1).expand(128, 5)
+        gamma = torch.exp(p_y_unorm - p_y_norm_e)  # (128, 5)
+
+        return dict(
+            mu=mu_ngc,
+            phi=phi,
+            gamma=gamma,
+            p_y_norm=p_y_norm,
+        )
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        n_obs: int = 1.0,
+    ):
+        # generative_outputs is a dict of the return value from `generative(...)`
+        # assume that `n_obs` is the number of training data points
+        p_y_norm = generative_outputs["p_y_norm"]
+
+        # compute Q
+        # gamma_fixed = torch.empty((None, self.n_labels))
+        # Q = -torch.einsum("nc,cn->", gamma_fixed, p_y_unorm)
```

Comment:
I'm assuming this or similar is correct for the components of Q that do not correspond to priors.

Recall we want to take the mean (or if einsum works, divide by number of cells in minibatch, and multiply by n_obs)

and then add prior terms.